java-checks/src/main/java/org/sonar/java/checks/regex/EmptyStringRepetitionCheck.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.util.Collections;
+import org.sonar.check.Rule;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.DisjunctionTree;
+import org.sonar.java.regex.ast.GroupTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
+import org.sonar.java.regex.ast.RepetitionTree;
+import org.sonar.java.regex.ast.SequenceTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5842")
+public class EmptyStringRepetitionCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Rework this part of the regex to not match the empty string.";
+
+  @Override
+  public void checkRegex(RegexParseResult regex, MethodInvocationTree mit) {
+    new Visitor().visit(regex);
+  }
+
+  private class Visitor extends RegexBaseVisitor {
+
+    @Override
+    public void visitRepetition(RepetitionTree tree) {
+      if (matchEmptyString(tree.getElement())) {
+        reportIssue(tree, MESSAGE, null, Collections.emptyList());
```

Comment:
The issue should either be reported on `tree.getElement()` instead of `tree` or the message should be changed. The way it is now, the message would imply that in `(x*)*` the whole expression should be changed to not match the empty string when in fact only the inner part (`(x*)`) should be changed to not match the empty string.